### PR TITLE
Synchronisation improvements

### LINKED ISSUE #67: ByteByByteSplit.java:43-45: Thread safety improvements....
The puzzle `32-0fde6650` from #32 has to be resolved: 

https://github.com/artipie/http/blob/5b955ad486f2a340e590e9d04142040d1ffe9fee/src/main/java/com/artipie/http/stream/ByteByByteSplit.java#L43-L45

The puzzle was created by @Sammers21 on 05-Mar-20. 

Estimate: 30 minutes,  role: DEV. 

If you have any technical questions, don't ask me, submit new tickets instead. The task will be \"done\" when the problem is fixed and the text of the puzzle is _removed_ from the source code. Here is more about [PDD](http://www.yegor256.com/2009/03/04/pdd.html) and [about me](http://www.yegor256.com/2017/04/05/pdd-in-action.html).

### PR DESCRIPTION
**Pull request by Sammers21**
This pull request resolves #67

### CODE REVIEW

**Comment by paulodamaso** [comment]
@g4s8 I'm idle, please assign this one to me

**Comment by 0crat** [comment]
This pull request [#82](https://github.com/artipie/http/issues/82) is assigned to @vatavuk[/z](https://www.0crat.com/u/vatavuk), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/aa66756d-2f2e-4fad-8b8b-8323da882307); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: src/main/java/com/artipie/http/stream/ByteByByteSplit.java [hunk 1]
```diff
@@ -144,13 +152,17 @@ public void subscribe(final Subscriber<? super Publisher<ByteBuffer>> sub) {
...
+                    synchronized (ByteByByteSplit.this.downSync) {
+                        ByteByByteSplit.this.downDemand.updateAndGet(operand -> operand + ask);
+                        ByteByByteSplit.this.upstream.get().get().request(ask);
+                    }
                 }
 
                 @Override
                 public void cancel() {
-                    throw new IllegalStateException("Cancel is not allowed");
+                    synchronized (ByteByByteSplit.this.downSync) {
```
Comment (Vatavuk) [inline src/main/java/com/artipie/http/stream/ByteByByteSplit.java]: @Sammers21 do we need synchronization on throwing exception?

Comment (Sammers21) [reply src/main/java/com/artipie/http/stream/ByteByByteSplit.java]: @Vatavuk, the rule is to acquire a specific lock for downstream operations, even if it does not look reasonable now.

Comment (Vatavuk) [reply src/main/java/com/artipie/http/stream/ByteByByteSplit.java]: @Sammers21 ok

**Comment by 0crat** [comment]
Job was finished in 12 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
Job audit: `ARC` didn't merge or close PR, performer @vatavuk[/z](https://www.0crat.com/u/vatavuk)[/z](https://www.0crat.com/u/vatavuk) didn't make CR comments

**Pull request merged**